crates/rattler_shell/src/activation.rs

```diff
@@ -459,21 +466,107 @@ impl<T: Shell + Clone> Activator<T> {
 
         script.set_path(path.as_slice(), variables.path_modification_behavior)?;
 
-        // deliberately not taking care of `CONDA_SHLVL` or any other complications at
-        // this point
+        // Get the current shell level
+        let shlvl = variables
+            .current_env
+            .get("PIXI_SHLVL")
+            .and_then(|s| s.parse::<i32>().ok())
+            .unwrap_or(0);
+
+        // Save original CONDA_PREFIX value if it exists
+        if let Some(existing_prefix) = variables.current_env.get("CONDA_PREFIX") {
+            script.set_env_var(
+                &format!("PIXI_ENV_SHLVL_{shlvl}_CONDA_PREFIX"),
+                existing_prefix,
+            )?;
+        }
+
+        // Set new CONDA_PREFIX
         script.set_env_var("CONDA_PREFIX", &self.target_prefix.to_string_lossy())?;
 
+        // For each environment variable that was set during activation
         for (key, value) in &self.env_vars {
+            // Save original value if it exists
+            if let Some(existing_value) = variables.current_env.get(key) {
+                script.set_env_var(&format!("PIXI_ENV_SHLVL_{shlvl}_{key}"), existing_value)?;
+            }
+            // Set new value
             script.set_env_var(key, value)?;
         }
 
+        // Set the new PIXI_SHLVL
+        script.set_env_var("PIXI_SHLVL", &(shlvl + 1).to_string())?;
+
         for activation_script in &self.activation_scripts {
             script.run_script(activation_script)?;
         }
 
         Ok(ActivationResult { script, path })
     }
 
+    /// Create a deactivation script for the environment.
+    /// This returns the deactivation script that unsets environment variables
+    /// and runs deactivation scripts.
+    pub fn deactivation(
+        &self,
+        variables: ActivationVariables,
+    ) -> Result<ActivationResult<T>, ActivationError> {
+        let mut script = ShellScript::new(self.shell_type.clone(), self.platform);
+
+        // Get the current PIXI shell level from passed environment variables
+        let current_pixi_shlvl = variables
+            .current_env
+            .get("PIXI_SHLVL")
+            .and_then(|s| s.parse::<i32>().ok());
+
+        match current_pixi_shlvl {
+            None | Some(0) => {
+                // Handle edge case: PIXI_SHLVL not set or invalid
+                script.echo("Warning: PIXI_SHLVL not set. This may indicate a broken workflow.")?;
+                script.echo(
+                    "Proceeding to unset pixi variables without restoring previous values.",
+                )?;
+
+                // Just unset without restoring
+                for (key, _) in &self.env_vars {
+                    script.unset_env_var(key)?;
+                }
+                script.unset_env_var("CONDA_PREFIX")?;
+                script.unset_env_var("PIXI_SHLVL")?;
+            }
+            Some(current_level) => {
+                let prev_shlvl = current_level - 1;
```

Comment:
here we need to get the current_level, if it 0 run `script.unset_env_var("PIXI_SHLVL")?;` and only then do the -1 level.
The match on `current_pixi_shlvl` should protect that we dont get the negative value